src/main/java/io/cryostat/diagnostic/Diagnostics.java

```diff
@@ -15,22 +15,185 @@
+    @Named(Producers.BASE64_URL)
+    Base64 base64Url;
+
+    @ConfigProperty(name = ConfigProperties.AWS_BUCKET_NAME_THREAD_DUMPS)
+    String bucket;
+
+    @ConfigProperty(name = ConfigProperties.STORAGE_PRESIGNED_DOWNLOADS_ENABLED)
+    boolean presignedDownloadsEnabled;
+
+    @ConfigProperty(name = ConfigProperties.STORAGE_EXT_URL)
+    Optional<String> externalStorageUrl;
+
+    @Inject EventBus bus;
+    @Inject DiagnosticsHelper helper;
+
+    @Path("targets/{targetId}/threaddump")
+    @RolesAllowed("write")
+    @Blocking
+    @POST
+    public String threadDump(
+            HttpServerResponse response, @RestPath long targetId, @RestQuery String format) {
+        log.tracev("Creating new thread dump request for target: {0}", targetId);
+        ThreadDumpRequest request =
+                new ThreadDumpRequest(
+                        UUID.randomUUID().toString(), Long.toString(targetId), format);
+        response.endHandler(
+                (e) -> bus.publish(LongRunningRequestGenerator.THREAD_DUMP_ADDRESS, request));
+        return request.id();
+    }
+
+    @Path("targets/{targetId}/threaddump")
+    @RolesAllowed("read")
+    @Blocking
+    @GET
+    public List<ThreadDump> getThreadDumps(@RestPath long targetId) {
+        log.tracev("Fetching thread dumps for target: {0}", targetId);
+        log.tracev("Thread dumps: {0}", helper.getThreadDumps(targetId));
+        log.tracev("Storage bucket: {0}", bucket);
+        return helper.getThreadDumps(targetId);
+    }
 
-    @Path("/gc")
+    @DELETE
+    @Blocking
+    @Path("targets/{targetId}/threaddump/{threadDumpId}")
+    @RolesAllowed("write")
+    public void deleteThreadDump(@RestPath String threadDumpId) {
+        try {
+            log.tracev("Deleting thread dump with ID: {0}", threadDumpId);
+            storage.headObject(
+                    HeadObjectRequest.builder().bucket(bucket).key(threadDumpId).build());
+        } catch (NoSuchKeyException e) {
+            throw new NotFoundException(e);
+        }
+        storage.deleteObject(
+                DeleteObjectRequest.builder().bucket(bucket).key(threadDumpId).build());
+    }
+
+    @Path("/threaddump/download/{encodedKey}")
+    @RolesAllowed("read")
+    @Blocking
+    @GET
+    public RestResponse<Object> handleStorageDownload(
+            @RestPath String encodedKey, @RestQuery String query) throws URISyntaxException {
+        Pair<String, String> decodedKey = helper.decodedKey(encodedKey);
+        var threadDumpId = decodedKey.getValue().strip();
+        log.tracev("Handling download Request for encodedKey: {0}", encodedKey);
+        log.tracev("Handling download Request for query: {0}", query);
+        log.tracev("Decoded key: {0}", decodedKey.toString());
+        log.tracev("UUID: {0}", threadDumpId);
+        log.tracev("Bucket: {0}", bucket);
+        storage.headObject(HeadObjectRequest.builder().bucket(bucket).key(threadDumpId).build())
+                .sdkHttpResponse();
```

Comment:
Does this `headObject` need to be wrapped in a try-catch for `NoSuchKeyException` like it is in `deleteThreadDump`? Or should the `.sdkHttpResponse()` value be used to check for the status code instead, maybe?